resources/assets/lib/user-preferences.ts

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the GNU Affero General Public License v3.0.
+// See the LICENCE file in the repository root for full licence text.
+
+import CurrentUser from 'interfaces/current-user';
+import { defaultUserPreferencesJson, UserPreferencesJson } from 'interfaces/current-user';
+import { route } from 'laroute';
+import { observable } from 'mobx';
+import { onErrorWithCallback } from 'utils/ajax';
+
+export default class UserPreferences {
+  @observable private current: UserPreferencesJson;
+  private updatingOptions = false;
+  private user?: CurrentUser;
+
+  constructor() {
+    this.current = Object.assign({}, defaultUserPreferencesJson, this.fromStorage());
+  }
+
+  get = <T extends keyof UserPreferencesJson>(key: T) => this.current[key];
+
+  set = <T extends keyof UserPreferencesJson>(key: T, value: UserPreferencesJson[T]) => {
```

Comment:
I don't think these been to be bound since they're always invoked directly instead of their reference being passed elsewhere? 🤔